lib/web_ui/dev/webdriver_browser.dart

```diff
@@ -0,0 +1,93 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:math';
+
+import 'package:image/image.dart';
+import 'package:webdriver/io.dart' show createDriver, WebDriver;
+
+import 'browser.dart';
+
+abstract class WebDriverBrowserEnvironment extends BrowserEnvironment {
+  late final Process _driverProcess;
+
+  Future<Process> spawnDriverProcess();
+  Uri get driverUri;
+
+  @override
+  Future<void> prepare() async {
+    _driverProcess = await spawnDriverProcess();
+
+    _driverProcess.stderr
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen((String error) {
+      print('[Webdriver] $error');
+    });
+
+    _driverProcess.stdout
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen((String log) {
+      print('[Webdriver] $log');
+    });
+  }
+
+  @override
+  Future<void> cleanup() async {
+    _driverProcess.kill();
+  }
+
+  @override
+  Future<Browser> launchBrowserInstance(Uri url, {bool debug = false}) async {
+    for(;;) {
+      try {
+        final WebDriver driver = await createDriver(
+          uri: driverUri, desired: <String, dynamic>{'browserName': packageTestRuntime.identifier});
+        return WebDriverBrowser(driver, url);
+      } on SocketException {
+        // Sometimes we may try to connect before the web driver port is ready. Retry.
+        print('Failed to connect to webdriver process. Retrying in 100 ms');
+        await Future<void>.delayed(const Duration(milliseconds: 100));
+      } catch (exception) {
+        print('Exception while creating webdriver: $exception');
+        rethrow;
+      }
+    }
```

Comment:
This could become a real infinite loop (e.g. if webdriver never starts for whatever reason). Since there's no timeout or max retry logic here, are you relying on the test suite's timeout to stop this loop? If that's the case, it would be nice to clarify that in a comment.